### PR TITLE
[xdoctest] reformat example code with google style in `paddle/io`

### PR DESCRIPTION
**Pull request by SigureMo**
### PR types

Others

### PR changes

Others

### Description

修改如下文件的示例代码为新的格式，并通过 `xdoctest` 检查：

- `python/paddle/io/dataloader/batch_sampler.py`
- `python/paddle/io/dataloader/dataset.py`
- `python/paddle/io/dataloader/sampler.py`
- `python/paddle/io/dataloader/worker.py`
- `python/paddle/io/reader.py`

预览：

- en: http://preview-paddle-pr-55732.paddle-docs-preview.paddlepaddle.org.cn/documentation/docs/en/api/paddle/io/BatchSampler_en.html
- cn: http://preview-paddle-pr-55732.paddle-docs-preview.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/io/BatchSampler_cn.html

@sunzhongkai588 @SigureMo @megemini

- #55629

PCard-66962

### CODE REVIEW

File: python/paddle/io/dataloader/dataset.py [hunk 1]
```diff
@@ -472,31 +494,34 @@ def random_split(dataset, lengths, generator=None):
...
+            2
+
+            >>> # output of the first subset
+            >>> for idx, v in enumerate(a_list[0]):
+            ...     print(idx, v)
+            >>> # doctest: +SKIP
+            0 1
+            1 3
+            2 9
+            >>> # doctest: -SKIP
```
Comment (sunzhongkai588) [inline python/paddle/io/dataloader/dataset.py]: just一些小疑问
- 为什么这部分要skip
- 我看dataset.py下有的api示例，输出给结果了，有的没给。是否要写结果，要不就根据原示例是否有结果，来决定

File: python/paddle/io/dataloader/batch_sampler.py [hunk 2]
```diff
@@ -339,27 +336,27 @@ def set_epoch(self, epoch):
...
+                ...         self.num_samples = num_samples
+                ...
+                ...     def __getitem__(self, idx):
+                ...         image = np.random.random([784]).astype('float32')
+                ...         label = np.random.randint(0, 9, (1, )).astype('int64')
+                ...         return image, label
+                ...
+                ...     def __len__(self):
+                ...         return self.num_samples
+                ...
```
Comment (sunzhongkai588) [inline python/paddle/io/dataloader/batch_sampler.py:362]: just一点小疑问，这部分如果去掉 `...` 是不是也没问题

File: python/paddle/io/reader.py [hunk 3]
```diff
@@ -321,68 +321,66 @@ class DataLoader:
...
+            ...     for i, (image, label) in enumerate(loader()):
+            ...         out = simple_net(image)
+            ...         loss = F.cross_entropy(out, label)
+            ...         avg_loss = paddle.mean(loss)
+            ...         avg_loss.backward()
+            ...         opt.minimize(avg_loss)
+            ...         simple_net.clear_gradients()
+            ...         print("Epoch {} batch {}: loss = {}".format(e, i, np.mean(loss.numpy())))
+
+    Notes:
```
Comment (sunzhongkai588) [inline python/paddle/io/reader.py:379]: 其实用 `.. note::` 也可以（狗头），anyway

**Review by sunzhongkai588 — DISMISSED** [review]
LGTM～
有一些小疑问辛苦一师傅解答一下
and 好像ci没过

File: python/paddle/io/dataloader/batch_sampler.py (hunk 2 above)
Comment (megemini) [reply python/paddle/io/dataloader/batch_sampler.py:362]: 确实是 ～ xdoctest 可以兼容只用 `>>> `，多行也不加 `... ` 提示。

但是，这只是 xdoctest 的特性，保不齐哪天不用 xdoctest 了，这样可能就挂了～ 

所以，个人建议，还是以 python 的 doctest 为最低兼容标准吧～

File: python/paddle/io/dataloader/batch_sampler.py [hunk 4]
```diff
@@ -58,40 +58,37 @@ class BatchSampler(Sampler):
...
+            ...     print(batch_indices)
+            ...
+            >>> # init with sampler
+            >>> sampler = RandomSampler(RandomDataset(100))
+            >>> bs = BatchSampler(sampler=sampler,
+            ...                     batch_size=8,
+            ...                     drop_last=True)
+            ...
+            >>> for batch_indices in bs:
+            ...     print(batch_indices)
```
Comment (megemini) [inline python/paddle/io/dataloader/batch_sampler.py]: 这部分的 print 要不要加个输出？

File: python/paddle/io/dataloader/dataset.py [hunk 5]
```diff
@@ -37,26 +37,25 @@ class Dataset:
...
+            ...         image = np.random.random([784]).astype('float32')
+            ...         label = np.random.randint(0, 9, (1, )).astype('int64')
+            ...         return image, label
+            ...
+            ...     def __len__(self):
+            ...         return self.num_samples
+            ...
+            >>> dataset = RandomDataset(10)
+            >>> for i in range(len(dataset)):
+            ...     print(dataset[i])
```
Comment (megemini) [inline python/paddle/io/dataloader/dataset.py]: 同上，要不要加个输出，或者加个 comment？

File: python/paddle/io/dataloader/dataset.py [hunk 6]
```diff
@@ -95,23 +94,23 @@ class IterableDataset(Dataset):
...
+            ...
+            ...     def __iter__(self):
+            ...         for i in range(self.num_samples):
+            ...             image = np.random.random([784]).astype('float32')
+            ...             label = np.random.randint(0, 9, (1, )).astype('int64')
+            ...             yield image, label
+            ...
+            >>> dataset = RandomDataset(10)
+            >>> for img, lbl in dataset:
+            ...     print(img, lbl)
```
Comment (megemini) [inline python/paddle/io/dataloader/dataset.py]: 同上，要不要加个输出？

另外，numpy 和 paddle 加个 seed，随即数量也改小点：

``` python
>>> import numpy as np
>>> import paddle
>>> from paddle.io import IterableDataset
>>> np.random.seed(2023)
>>> paddle.seed(2023)
>>> # define a random dataset
>>> class RandomDataset(IterableDataset):
...     def __init__(self, num_samples):
...         self.num_samples = num_samples
...
...     def __iter__(self):
...         for i in range(self.num_samples):
...             image = np.random.random([3]).astype('float32')
...             label = np.random.randint(0, 9, (1, )).astype('int64')
...             yield image, label
...
>>> dataset = RandomDataset(3)
>>> for img, lbl in dataset:
...     print(img, lbl)
[0.3219883  0.89042246 0.5880523 ] [3]
[0.04382154 0.766739   0.33634686] [1]
[0.7272747  0.52438736 0.5449352 ] [8]
```

看看这样行不？之前的也是～

File: python/paddle/io/dataloader/dataset.py [hunk 7]
```diff
@@ -249,22 +274,21 @@ class TensorDataset(Dataset):
...
+            >>> label = paddle.to_tensor(label_np)
 
-            for i in range(len(dataset)):
-                input, label = dataset[i]
-                print(input, label)
+            >>> dataset = TensorDataset([input, label])
 
+            >>> for i in range(len(dataset)):
+            ...     input, label = dataset[i]
+            ...     print(input, label)
```
Comment (megemini) [inline python/paddle/io/dataloader/dataset.py]: 同上～

File: python/paddle/io/dataloader/dataset.py [hunk 8]
```diff
@@ -309,32 +333,31 @@ class ComposeDataset(Dataset):
...
+            ...     def __len__(self):
+            ...         return self.num_samples
+            ...
+            >>> dataset = ComposeDataset([RandomDataset(10), RandomDataset(10)])
+            >>> for i in range(len(dataset)):
+            ...     image1, label1, image2, label2 = dataset[i]
+            ...     print(image1)
+            ...     print(label1)
+            ...     print(image2)
+            ...     print(label2)
```
Comment (megemini) [inline python/paddle/io/dataloader/dataset.py]: 同上～

File: python/paddle/io/dataloader/dataset.py [hunk 9]
```diff
@@ -379,26 +402,25 @@ class ChainDataset(IterableDataset):
...
+            ...
+            ...     def __iter__(self):
+            ...         for i in range(10):
+            ...             image = np.random.random([32]).astype('float32')
+            ...             label = np.random.randint(0, 9, (1, )).astype('int64')
+            ...             yield image, label
+            ...
+            >>> dataset = ChainDataset([RandomDataset(10), RandomDataset(10)])
+            >>> for image, label in iter(dataset):
+            ...     print(image, label)
```
Comment (megemini) [inline python/paddle/io/dataloader/dataset.py]: 同上～

File: python/paddle/io/dataloader/dataset.py [hunk 10]
```diff
@@ -472,31 +494,34 @@ def random_split(dataset, lengths, generator=None):
...
+            >>> import paddle
+            >>> from paddle.io import random_split
+
+            >>> a_list = paddle.io.random_split(range(10), [3, 7])
+            >>> print(len(a_list))
+            2
+
+            >>> # output of the first subset
+            >>> for idx, v in enumerate(a_list[0]):
+            ...     print(idx, v)
```
Comment (megemini) [inline python/paddle/io/dataloader/dataset.py]: 加个 seed 就行，应该不用skip～

File: python/paddle/io/dataloader/dataset.py [hunk 11]
```diff
@@ -472,31 +494,34 @@ def random_split(dataset, lengths, generator=None):
...
+            >>> # doctest: +SKIP
+            0 5
+            1 7
+            2 8
+            3 6
+            4 0
+            5 2
+            6 4
+            >>> # doctest: -SKIP
```
Comment (megemini) [inline python/paddle/io/dataloader/dataset.py]: 同上～

File: python/paddle/io/dataloader/sampler.py [hunk 12]
```diff
@@ -44,34 +44,39 @@ class Sampler:
...
+            ...
+            >>> sampler = MySampler(data_source=RandomDataset(100))
+
+            >>> for index in sampler:
+            ...     print(index)
+            0
+            1
+            2
+            ...
+            99
```
Comment (megemini) [inline python/paddle/io/dataloader/sampler.py:79]: 这个好！我加到 issue 的描述里面做个范例，后面更新文档也用的上～

File: python/paddle/io/dataloader/sampler.py [hunk 13]
```diff
@@ -160,25 +170,24 @@ class RandomSampler(Sampler):
...
+            ...         label = np.random.randint(0, 9, (1, )).astype('int64')
+            ...         return image, label
+            ...
+            ...     def __len__(self):
+            ...         return self.num_samples
+            ...
+            >>> sampler = RandomSampler(data_source=RandomDataset(100))
+
+            >>> for index in sampler:
+            ...     print(index)
```
Comment (megemini) [inline python/paddle/io/dataloader/sampler.py:192]: 同上，加 seed 加 输出～

File: python/paddle/io/dataloader/sampler.py [hunk 14]
```diff
@@ -288,14 +297,14 @@ class WeightedRandomSampler(Sampler):
...
+            >>> from paddle.io import WeightedRandomSampler
 
-            for index in sampler:
-                print(index)
+            >>> sampler = WeightedRandomSampler(weights=[0.1, 0.3, 0.5, 0.7, 0.2],
+            ...                                 num_samples=5,
+            ...                                 replacement=True)
+            ...
+            >>> for index in sampler:
+            ...     print(index)
```
Comment (megemini) [inline python/paddle/io/dataloader/sampler.py:317]: 同上，加 seed 加 输出～

File: python/paddle/io/reader.py [hunk 15]
```diff
@@ -321,68 +321,66 @@ class DataLoader:
...
+            ...
+            >>> for e in range(EPOCH_NUM):
+            ...     for i, (image, label) in enumerate(loader()):
+            ...         out = simple_net(image)
+            ...         loss = F.cross_entropy(out, label)
+            ...         avg_loss = paddle.mean(loss)
+            ...         avg_loss.backward()
+            ...         opt.minimize(avg_loss)
+            ...         simple_net.clear_gradients()
+            ...         print("Epoch {} batch {}: loss = {}".format(e, i, np.mean(loss.numpy())))
```
Comment (megemini) [inline python/paddle/io/reader.py:377]: 这个地方算是一个相对完整的 case，输出加不加我觉得都行～

File: python/paddle/io/dataloader/batch_sampler.py [hunk 16]
```diff
@@ -203,29 +200,29 @@ class DistributedBatchSampler(BatchSampler):
...
+            ...
+            ...     def __len__(self):
+            ...         return self.num_samples
+            ...
+            >>> dataset = RandomDataset(100)
+            >>> sampler = DistributedBatchSampler(dataset, batch_size=64)
+
+            >>> for data in sampler:
+            ...     # do something
+            ...     break
```
Comment (megemini) [inline python/paddle/io/dataloader/batch_sampler.py:232]: 或者其他的几个 print 的地方都改成这样也不错～

File: python/paddle/io/dataloader/dataset.py [hunk 17]
```diff
@@ -125,87 +124,113 @@ def __iter__(self):
...
+            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
+                [[4]])
+            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
+                [[5]])
+            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
+                [[6]])
+            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
+                [[7]])
+            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
+                [[8]])
```
Comment (megemini) [inline python/paddle/io/dataloader/dataset.py]: 这里 num_workers=2，输出行为不可控吧？是不是加个 skip？下面的几个也是～ 我看日志里面 fail 的都是这样～

File: python/paddle/io/dataloader/dataset.py (hunk 1 above)
Comment (SigureMo) [reply python/paddle/io/dataloader/dataset.py]: https://github.com/PaddlePaddle/Paddle/pull/55849#discussion_r1280447574

都听顺师傅的（啊不是，其实我本来也是这么想的）

File: python/paddle/io/reader.py (hunk 3 above)
Comment (SigureMo) [reply python/paddle/io/reader.py:379]: 如果没记错的话，这是当年 @Ligoml 她老人家教导的，要改成这样的 [doge]

当然也有可能是我记错了 [doge][doge][doge]

File: python/paddle/io/dataloader/batch_sampler.py (hunk 16 above)
Comment (SigureMo) [reply python/paddle/io/dataloader/batch_sampler.py:232]: 修改了一版，整体对于 image 的都采用这个策略，因为大多数不需要 print，直接加注释 do something

其他的该加输出的也都加了

File: python/paddle/io/dataloader/worker.py [hunk 18]
```diff
@@ -94,51 +94,64 @@ def get_worker_info():
...
+            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
+            [[4]])
+            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
+            [[5]])
+            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
+            [[6]])
+            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
+            [[7]])
+            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
+            [[8]])
```
Comment (sunzhongkai588) [inline python/paddle/io/dataloader/worker.py]: 试了下，这儿的输出好像是固定的？
```suggestion
            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
            [[2]])
            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
            [[6]])
            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
            [[3]])
            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
            [[7]])
            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
            [[4]])
            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
            [[8]])
            Tensor(shape=[1, 1], dtype=int64, place=Place(cpu), stop_gradient=True,
            [[5]])
```

File: python/paddle/io/dataloader/dataset.py [hunk 19]
```diff
@@ -472,31 +495,31 @@ def random_split(dataset, lengths, generator=None):
...
+            >>> # output of the second subset
+            >>> for idx, v in enumerate(a_list[1]):
+            ...     print(idx, v)
+            0 9
+            1 6
+            2 3
+            3 4
+            4 1
+            5 0
+            6 7
```
Comment (sunzhongkai588) [inline python/paddle/io/dataloader/dataset.py]: 这是不是api本身有问题..？我在aistudio上的gpu和cpu的环境分别试了下，两个结果都不一样（即使指定`seed`和`cpuplace`）

File: python/paddle/io/dataloader/worker.py (hunk 18 above)
Comment (SigureMo) [reply python/paddle/io/dataloader/worker.py]: 离谱，我本地可稳定的 2345678 了

image

开发机和 CI 一样

image

按照 CI 来吧

File: python/paddle/io/dataloader/dataset.py (hunk 19 above)
Comment (SigureMo) [reply python/paddle/io/dataloader/dataset.py]: image

image

对哦，为啥

**Review by sunzhongkai588 — APPROVED** [review]
LGTM
